cornice/schemas.py

```diff
@@ -1,180 +1,78 @@
 # This Source Code Form is subject to the terms of the Mozilla Public
 # License, v. 2.0. If a copy of the MPL was not distributed with this file,
 # You can obtain one at http://mozilla.org/MPL/2.0/.
-import webob.multidict
+from cornice.util import extract_request_data
 
-from pyramid.path import DottedNameResolver
-from cornice.util import to_list, extract_request_data
 
-
-class SchemaError(Exception):
+try:
+    import colander
+except ImportError:
     pass
-
-
-class CorniceSchema(object):
-    """Defines a cornice schema"""
-
-    def __init__(self, _colander_schema, bind_request=True):
-        self._colander_schema = _colander_schema
-        self._colander_schema_runtime = None
-        self._bind_request = bind_request
-
-    @property
-    def colander_schema(self):
-        if not self._colander_schema_runtime:
-            schema = self._colander_schema
-            schema = DottedNameResolver(__name__).maybe_resolve(schema)
-            if callable(schema):
-                schema = schema()
-            self._colander_schema_runtime = schema
-        return self._colander_schema_runtime
-
-    def bind_attributes(self, request=None):
-        schema = self.colander_schema
-        if request and self._bind_request:
-            schema = schema.bind(request=request)
-        return schema.children
-
-    def get_attributes(self, location=("body", "header", "querystring"),
-                       required=(True, False),
-                       request=None):
-        """Return a list of attributes that match the given criteria.
-
-        By default, if nothing is specified, it will return all the attributes,
-        without filtering anything.
+else:
+
+    class StrictMappingSchema(colander.MappingSchema):
+        @staticmethod
+        def schema_type():
+            return colander.Mapping(unknown='raise')
+
+    class CorniceSchema(colander.MappingSchema):
+        querystring = StrictMappingSchema()
+        headers = StrictMappingSchema()
+        body = StrictMappingSchema()
+        path = StrictMappingSchema()
```

Comment:
Another issue with the old system is everything ends up in `request.validated` so, for example, if you have a "body" element and a "querystring" with the same name they'll conflict.  The new one ends up with `request.validated['body']` and `request.validated['querystring']`.  I wrote a `from_colander()` method, but the result in request.validated will be completely different then before.

I don't think there's a way to support a smooth transition from the old...  So, I think it's best to fail early and hard so people are forced to face the change directly.